README.md
# coWin-vaccine
This project is developed to book vaccine slots automatically that comes on coWin website. 

## Updates
### Important: 
- This is a just proof of concept project.
- Make sure that all the beneficiaries selected are supposed to be of same age and to get the same vaccine and dose. 
- There is no option to register new mobile or add beneficiaries. This can be used only after beneficiary has been added through the official app/site.
- And finally, I know code quality isn't great. Suggestions are welcome.

### Usage:

Use **Python 3.7** and install all the dependencies with:
```
pip install -r requirements.txt
```

Then, run the script file as show below:
```
python src\main.py
```

If you already have a bearer token, you can also use:
```
python src\main.py --token=YOUR-TOKEN-HERE
```

### Third-Party Package Dependency:
- ```PrettyTable``` : For displaying data in tabular format.
- ```requests``` : For making GET and POST requests to the API.
- ```decouple``` : For picking values from an env file
- ```svglib``` : For rendering the captcha

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 1.0.1 (2021-06-13)


### Bug Fixes

* condition is changed for fetching the slots ([859062a](https://github.com/mnkmalpani/coWin-vaccine/commit/859062a9c2828116dea0fdb7763f9a936ae47922))


test.py
from src.captcha import captcha_builder

response = {
    "captcha": "<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"150\" height=\"50\" viewBox=\"0,0,150,50\"><path fill=\"#111\" d=\"M99.95 29.45L100.03 29.53L100.01 29.50Q98.88 28.80 98.24 28.83L98.08 28.68L98.15 28.75Q96.77 28.81 95.72 29.82L95.74 29.84L95.80 29.90Q94.76 30.91 94.87 32.28L94.72 32.14L94.89 32.30Q95.10 36.55 94.22 41.08L94.35 41.20L94.18 41.03Q91.98 41.50 90.88 42.03L90.86 42.01L90.97 42.12Q92.41 37.89 92.18 33.43L92.22 33.47L92.13 33.38Q92.03 29.03 90.32 24.95L90.33 24.96L90.22 24.85Q91.43 25.60 93.71 26.25L93.61 26.15L94.28 28.95L94.22 28.89Q95.03 26.12 99.29 26.12L99.13 25.96L99.11 25.94Q99.63 26.00 99.93 26.00L99.81 25.87L99.79 25.86Q101.12 25.90 102.38 26.58L102.36 26.56L102.34 26.54Q101.15 28.02 100.01 29.50ZM101.66 31.53L101.75 31.47L101.78 31.50Q102.59 30.25 104.26 27.82L104.34 27.90L104.26 27.81Q103.81 27.48 102.29 27.21L102.29 27.21L102.31 27.24Q102.62 26.90 103.08 26.29L103.10 26.31L103.10 26.31Q101.19 25.39 99.25 25.58L99.18 25.51L99.31 25.64Q97.28 25.75 96.56 26.01L96.48 25.94L96.54 26.00Q95.22 26.50 94.42 27.60L94.31 27.50L94.23 26.77L94.03 25.92L93.95 25.84Q91.56 25.43 89.58 24.09L89.61 24.13L89.51 24.03Q91.67 28.43 91.90 33.31L92.02 33.43L91.89 33.30Q92.20 38.29 90.45 42.82L90.49 42.86L90.35 42.73Q91.50 42.31 92.38 42.01L92.32 41.95L92.30 41.93Q92.04 43.04 91.55 44.11L91.65 44.21L91.71 44.27Q94.38 43.25 96.51 43.14L96.57 43.19L96.87 38.78L96.73 38.63Q96.97 36.52 96.97 34.27L96.88 34.18L96.95 34.25Q97.00 33.12 97.91 32.15L97.92 32.15L97.86 32.10Q98.70 31.05 99.85 30.98L99.98 31.11L99.91 31.04Q100.48 30.85 101.21 31.54L101.18 31.51L101.56 31.44ZM96.54 29.61L96.61 29.60L96.59 29.70L96.51 29.59Z\"/><path d=\"M19 43 C85 47,72 45,131 49\" stroke=\"#888\" fill=\"none\"/><path fill=\"#111\" d=\"M121.22 41.33L121.20 41.31L121.25 41.36Q118.30 41.46 117.05 40.93L117.17 41.05L117.03 40.91Q115.17 40.19 115.24 37.72L115.34 37.81L115.27 37.74Q115.38 34.31 115.38 34.88L115.37 34.87L115.32 34.82Q115.39 29.56 112.53 24.73L112.56 24.75L112.60 24.79Q114.59 25.76 116.57 26.14L116.57 26.14L116.41 25.98Q118.16 30.62 118.05 35.12L118.13 35.19L118.05 35.11Q117.93 36.63 118.76 37.85L118.76 37.84L118.79 37.87Q119.69 39.19 121.21 39.00L121.18 38.97L121.62 38.91L121.69 38.98Q123.02 38.90 123.84 38.14L123.73 38.04L123.78 38.09Q124.59 37.31 124.59 36.36L124.54 36.32L124.50 36.27Q124.60 36.15 124.56 35.96L124.62 36.01L124.45 35.85Q124.40 30.62 125.54 26.28L125.64 26.37L125.56 26.29Q126.58 26.21 127.57 25.98L127.63 26.04L129.55 25.45L129.64 25.54Q127.03 30.51 127.34 36.06L127.29 36.02L127.46 36.19Q127.54 39.16 125.82 40.30L125.95 40.42L125.86 40.34Q124.64 41.29 121.33 41.44ZM123.61 43.80L123.49 43.68L123.57 43.76Q127.05 43.92 128.87 43.16L128.84 43.13L128.97 43.26Q129.83 42.45 129.83 41.04L129.71 40.91L129.77 40.98Q129.72 40.01 129.45 38.45L129.59 38.58L129.46 38.45Q129.07 35.40 129.45 32.39L129.39 32.33L129.46 32.41Q129.78 29.18 131.19 26.44L131.22 26.48L131.18 26.43Q130.65 26.85 129.31 27.27L129.17 27.12L129.21 27.16Q129.49 26.60 129.79 26.03L129.85 26.09L130.35 24.88L130.44 24.97Q128.49 25.61 125.33 26.10L125.35 26.13L125.16 25.93Q124.08 30.87 124.24 35.97L124.28 36.02L124.22 35.96Q124.31 38.07 121.65 38.56L121.52 38.44L121.35 38.72L121.11 38.63L120.19 38.51L120.10 38.42Q120.01 37.53 120.05 37.04L120.11 37.10L120.07 37.06Q120.27 32.54 119.59 28.13L119.68 28.22L118.53 27.99L118.48 27.93Q118.02 27.97 117.49 27.85L117.43 27.80L117.57 27.93Q117.16 26.50 116.89 25.74L116.76 25.60L116.74 25.59Q113.96 25.17 111.72 23.87L111.79 23.95L111.85 24.01Q114.03 27.55 114.75 30.94L114.74 30.93L114.66 30.86Q115.09 33.00 114.94 36.50L115.10 36.65L115.09 36.64Q115.09 36.76 115.01 37.87L115.02 37.87L114.99 37.84Q114.81 40.18 116.18 41.13L116.28 41.23L116.23 41.18Q116.49 43.30 123.50 43.68Z\"/><path d=\"M12 25 C68 8,89 17,136 18\" stroke=\"#222\" fill=\"none\"/><path fill=\"#444\" d=\"M80.04 23.37L80.03 23.36L79.91 23.24Q78.20 19.17 72.00 19.36L72.09 19.46L71.97 19.33Q67.62 19.51 65.64 20.96L65.56 20.88L65.60 20.92Q63.26 22.73 63.26 26.84L63.19 26.77L63.15 28.60L63.25 28.70Q63.19 37.47 72.21 37.47L72.30 37.56L72.18 37.44Q74.40 37.60 75.89 36.80L75.78 36.70L75.83 36.74Q77.71 35.77 77.56 33.91L77.50 33.85L77.52 33.87Q77.31 31.76 73.62 30.85L73.70 30.92L71.75 30.92L71.66 30.83Q70.66 30.78 69.71 30.82L69.73 30.84L69.79 30.90Q69.72 29.87 69.68 27.97L69.73 28.02L69.62 27.91Q77.40 28.04 81.59 27.70L81.47 27.58L81.51 27.62Q81.45 28.44 81.45 29.24L81.46 29.24L81.43 29.21Q81.50 30.09 81.54 30.89L81.56 30.91L81.46 30.80Q80.45 30.90 80.45 30.90L80.45 30.90L80.52 30.97Q79.05 30.86 79.43 30.86L79.43 30.87L79.40 30.84Q80.59 32.57 80.98 34.55L81.00 34.57L80.99 34.56Q80.97 35.15 80.97 35.61L80.94 35.58L81.07 35.71Q80.93 40.21 72.14 40.21L72.25 40.33L72.21 40.29Q65.16 40.36 62.57 38.26L62.58 38.27L62.52 38.21Q60.50 35.81 60.50 30.74L60.50 30.74L60.63 30.88Q60.55 29.72 60.39 27.02L60.42 27.05L60.32 26.94Q60.25 24.67 60.29 23.30L60.23 23.24L60.18 23.19Q60.33 20.37 61.70 18.77L61.69 18.76L61.77 18.84Q64.36 16.82 69.23 16.63L69.27 16.68L71.94 16.60L71.93 16.59Q75.01 16.48 78.44 16.90L78.39 16.84L78.33 16.79Q82.18 17.36 83.13 20.60L83.18 20.64L82.99 20.45Q82.05 21.46 79.96 23.29ZM81.38 25.51L81.39 25.40L81.39 25.41Q82.59 24.32 84.46 22.34L84.61 22.49L84.53 22.42Q84.28 21.51 83.29 19.80L83.25 19.77L83.29 19.81Q81.98 17.01 78.17 16.36L78.15 16.34L78.21 16.39Q76.24 16.45 71.94 16.15L71.90 16.10L69.20 16.23L69.13 16.15Q63.96 16.35 61.37 18.33L61.33 18.29L61.46 18.42Q60.01 19.97 59.90 23.06L59.88 23.04L59.94 23.10Q59.95 22.66 60.18 28.14L60.18 28.14L60.20 28.16Q60.12 29.41 60.19 32.15L60.27 32.22L60.30 32.25Q60.36 36.59 62.12 38.53L62.12 38.53L62.14 38.55Q62.49 39.36 63.37 40.24L63.47 40.34L63.38 40.25Q66.39 42.23 71.91 42.57L71.95 42.61L74.72 42.68L74.69 42.65Q76.29 42.77 78.31 42.66L78.34 42.68L78.38 42.73Q79.85 42.67 81.94 42.14L82.00 42.19L81.88 42.07Q83.92 41.67 83.88 39.35L83.71 39.18L83.89 39.36Q83.79 38.00 82.72 35.22L82.65 35.15L82.73 35.23Q82.47 34.36 81.94 33.56L81.97 33.60L81.88 33.51Q82.53 33.58 83.71 33.77L83.63 33.69L83.78 33.85Q83.39 30.14 83.43 29.46L83.41 29.44L83.50 29.53Q82.95 29.36 82.53 29.32L82.56 29.35L82.66 29.46Q82.25 29.39 81.87 29.46L81.91 29.50L81.94 29.53Q81.87 28.36 81.95 27.29L81.87 27.21L81.88 27.23Q75.67 27.79 69.39 27.68L69.34 27.63L69.41 27.70Q69.36 28.53 69.36 29.44L69.50 29.58L69.49 31.32L69.36 31.19Q69.89 31.34 70.31 31.30L70.22 31.21L70.20 31.20Q70.63 31.21 71.09 31.21L71.06 31.18L71.15 31.27Q71.08 31.65 71.12 32.15L71.11 32.14L71.14 32.17Q71.12 32.61 71.08 33.10L71.01 33.03L71.04 33.06Q72.05 33.01 73.08 32.97L73.16 33.05L73.16 33.05Q74.20 33.06 75.22 33.10L75.14 33.01L75.28 33.15Q76.30 33.37 77.17 33.75L77.18 33.76L77.10 33.72L77.07 33.68Q77.34 35.33 75.67 36.24L75.73 36.30L75.67 36.24Q74.00 37.16 72.29 37.16L72.33 37.20L72.30 37.17Q68.34 37.06 65.90 35.31L66.06 35.46L66.02 35.42Q65.03 33.37 65.19 30.82L65.27 30.91L65.13 30.77Q65.49 26.03 67.34 24.03L67.28 23.97L67.41 24.10Q69.19 22.04 73.87 21.73L73.80 21.66L73.78 21.63Q77.55 21.53 79.49 23.31L79.45 23.28L79.54 23.36Q79.51 23.37 79.62 23.56L79.66 23.60L79.78 23.61L79.80 23.62Q80.43 24.37 81.00 25.47L80.92 25.39L81.26 25.39Z\"/><path fill=\"#444\" d=\"M15.92 49.57L15.89 49.54L15.89 49.53Q18.97 45.84 21.87 39.63L21.83 39.60L21.74 39.51Q18.73 33.08 12.64 25.01L12.53 24.89L12.49 24.86Q14.43 26.03 17.21 26.68L17.17 26.64L17.32 26.79Q20.87 31.49 23.31 36.70L23.15 36.55L23.17 36.56Q25.36 31.29 28.44 26.87L28.50 26.93L28.42 26.85Q30.58 26.46 32.68 25.66L32.69 25.67L32.84 25.83Q29.39 29.61 25.93 37.04L25.93 37.03L23.10 43.04L23.01 42.95Q21.65 46.16 19.90 48.78L19.78 48.67L19.83 48.71Q17.47 49.13 16.02 49.66ZM22.17 50.98L22.21 51.02L22.19 51.00Q23.89 48.09 25.34 44.82L25.28 44.76L28.03 38.46L28.12 38.55Q31.58 30.46 34.89 26.47L34.85 26.43L34.96 26.53Q33.62 27.22 31.72 27.90L31.64 27.82L32.58 26.33L32.65 26.39Q33.11 25.60 33.72 24.91L33.90 25.09L33.84 25.04Q31.22 26.15 28.36 26.60L28.26 26.50L28.35 26.59Q25.60 30.50 23.66 34.96L23.67 34.97L23.52 34.82Q21.88 31.51 20.09 28.69L20.18 28.77L20.26 28.86Q20.04 28.83 19.68 28.79L19.55 28.66L19.66 28.77Q19.28 28.71 19.09 28.71L18.98 28.60L19.11 28.73Q18.24 27.37 17.29 26.23L17.30 26.24L17.40 26.34Q14.21 25.66 11.47 23.94L11.51 23.99L11.45 23.93Q17.64 31.87 21.34 39.49L21.39 39.54L21.29 39.44Q19.16 44.54 15.09 50.18L15.06 50.15L15.20 50.29Q15.89 49.88 16.69 49.72L16.77 49.81L18.31 49.36L18.48 49.53Q17.72 50.11 16.54 51.75L16.60 51.80L16.56 51.76Q18.93 51.24 22.20 51.01Z\"/><path fill=\"#444\" d=\"M44.30 26.78L44.44 26.92L44.40 26.88Q46.14 26.98 47.97 26.98L48.03 27.04L47.96 26.97Q49.72 26.94 51.54 26.79L51.68 26.93L51.60 26.85Q51.46 27.47 51.46 28.19L51.48 28.21L51.45 29.52L51.51 29.57Q49.12 29.55 44.36 29.66L44.28 29.57L44.40 29.70Q44.35 35.58 43.20 40.57L43.22 40.59L43.20 40.56Q40.96 41.22 39.48 42.02L39.56 42.10L39.53 42.07Q41.86 35.27 41.60 28.15L41.55 28.10L41.52 28.07Q41.39 21.05 38.72 14.39L38.60 14.27L38.72 14.39Q42.28 16.81 47.04 17.04L47.15 17.15L47.07 17.06Q51.70 17.20 55.85 15.45L55.97 15.57L55.88 15.48Q55.60 16.31 55.37 17.11L55.48 17.21L54.89 18.72L54.87 18.69Q52.29 19.92 49.32 20.07L49.22 19.98L49.13 19.89Q46.34 20.18 43.48 19.46L43.34 19.31L43.38 19.35Q44.08 22.75 44.27 26.75ZM56.53 14.80L56.61 14.88L56.53 14.80Q51.95 16.95 47.07 16.65L47.07 16.65L47.06 16.63Q41.77 16.37 38.04 13.52L37.97 13.45L38.07 13.54Q40.80 20.35 41.10 27.96L41.16 28.02L41.09 27.95Q41.57 36.05 39.06 42.82L38.99 42.75L38.96 42.73Q39.60 42.29 40.97 41.65L40.96 41.64L40.97 41.64Q40.79 42.42 40.22 43.94L40.31 44.03L40.36 44.09Q42.59 42.92 45.48 42.43L45.39 42.34L45.46 42.41Q46.05 37.51 46.24 31.80L46.25 31.81L46.28 31.84Q48.07 31.81 49.90 31.81L49.96 31.87L49.86 31.77Q51.77 31.89 53.60 32.08L53.45 31.93L53.57 32.06Q53.33 30.98 53.33 30.06L53.49 30.23L53.31 28.14L53.48 28.31Q52.69 28.39 51.85 28.39L51.75 28.29L51.82 28.36Q51.88 27.55 51.96 26.52L51.82 26.38L51.84 26.40Q50.53 26.65 49.12 26.65L49.11 26.65L49.19 26.72Q47.79 26.73 46.42 26.69L46.30 26.57L46.25 24.39L46.22 24.37Q46.16 23.24 46.05 22.17L46.06 22.19L46.16 22.28Q47.06 22.27 48.05 22.27L48.02 22.24L48.02 22.24Q52.87 22.33 56.49 20.43L56.53 20.47L56.48 20.42Q56.90 18.52 57.81 15.89L57.78 15.86L57.94 16.02Q56.60 16.81 55.80 17.15L55.76 17.12L55.75 17.11Q56.10 16.42 56.59 14.86Z\"/><path d=\"M2 20 C86 5,83 30,132 40\" stroke=\"#555\" fill=\"none\"/></svg>"
}


if __name__ == '__main__':
    captcha = captcha_builder(response)
    print(captcha)



src/main.py
import sys
from datetime import datetime
import argparse
from utils import *
import os

header_otp = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36"
}

header = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36"
    , 'content-type': 'application/json'
    , 'origin': 'https://selfregistration.cowin.gov.in'
    , 'sec-fetch-site': 'cross-site'
    , 'sec-fetch-mode': 'cors'
    , 'sec-fetch-dest': 'empty'
    , 'referer': 'https://selfregistration.cowin.gov.in/'
}

beneficiaries_data = {}

# mobile = config('MOBILE')
secret = "U2FsdGVkX1/VsmHZHLbdwntV6fMy5vTmAZhQtNlj00zdmonoostJjETavz9NKf578AFc3y1bgqAvLExdg48bRA=="
TOKEN_VALIDITY = 840  # taking 1 min buffer

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--token', help='Pass token directly')
    parser.add_argument('--auto', help='otp automatic reading capability')
    args = parser.parse_args()

    if not args.token:
        mobile = int(input("Please enter your mobile number without extension code: ").strip())
        if not args.auto:
            token = generate_and_validate_otp(mobile, header_otp, secret, False)
        else:
            print("Auto otp fetch and verification enabled")
            token = generate_and_validate_otp(mobile, header_otp, secret, True)
    else:
        token = args.token

    authentication = f"Bearer {token}"
    token_generated_time = datetime.now()

    # fetch beneficiaries data and print them
    beneficiaries_response = beneficiaries(authentication, header)
    beneficiaries_data = showBeneficiariesDetails(beneficiaries_response)

    # selection of the beneficiaries
    selection = [x.strip() for x in input("Select the person to get vacinated with comma separated values: ").split(',')]
    beneficiary_reference_ids = []
    for index in selection:
        beneficiary_reference_ids.append(beneficiaries_data.get(index).get('beneficiary_reference_id'))
    # select state & district
    states_response = getStates(authentication, header)
    showStates(states_response)
    state = input("Select the state: ")
    districts_response = getDistricts(state, header)
    showDistricts(districts_response)
    districts = [int(x.strip()) for x in input("Select the district with comma separated values: ").split(',')]

    # select date
    date = getDateFromUser()

    # get list of conditions
    fee_type, fee_type_flag, vaccine, vaccine_flag, dose_number_preference = getUserConditions()

    #get the min age
    ages = []
    for index in selection:
        ages.append(beneficiaries_data.get(index).get('age'))
    min_age_limit = 18 if max(ages) < 45 else 45
    print(f"min_age_limit: {min_age_limit}")

    #min slot required
    mini_slot = len(selection)

    while (True):
        # find session
        sessions_dict = getSessionsByDistrict(districts, date, header)
        sessions_list = sessions_dict.get("sessions", [])

        for item in sessions_list:
            # create condition list
            condition = [item.get('min_age_limit') == min_age_limit,
                         item.get('available_capacity') >= mini_slot]
            if dose_number_preference == 2:
                condition.append(item.get('available_capacity_dose2') >= mini_slot)
            else:
                condition.append(item.get('available_capacity_dose1') >= mini_slot)
            if fee_type_flag:
                condition.append(fee_type == item.get('fee_type'))
            if vaccine_flag:
                condition.append(vaccine.upper() == item.get('vaccine'))

            if all(condition):
                for slot in item.get("slots"):
                    print("Slot available: ", item)
                    header["Authorization"] = authentication
                    response = book_slot(header=header, mini_slot=mini_slot, session_id=item.get('session_id'),
                                         slot=slot, beneficiary_reference_ids=beneficiary_reference_ids, center_id=item.get("center_id"))
                    print("Booking response: ", response)
                    if response.status_code != 200:
                        print("Somthing went wrong, checking next available slot")
                    else:
                        print('##############    BOOKED!  ############################    BOOKED!  ##############')
                        print('\nPress any key thrice to exit program.')
                        os.system("pause")
                        os.system("pause")
                        os.system("pause")
                        sys.exit()

        print("No slots found, checking again. Token validity time remaning in seconds: ",
              (TOKEN_VALIDITY - (datetime.now() - token_generated_time).seconds))
        # check validity
        if (datetime.now() - token_generated_time).seconds > TOKEN_VALIDITY:
            print("Token is Invalid")
            if not args.auto:
                token = generate_and_validate_otp(mobile, header_otp, secret, False)
            else:
                print("Automation enabled")
                token = generate_and_validate_otp(mobile, header_otp, secret, True)
            authentication = f"Bearer {token}"
            if not check_token_status(authentication=authentication, header=header):
                print("token Not valid, try again")
                sys.exit(1)
            token_generated_time = datetime.now()


src/captcha.py
from svglib.svglib import svg2rlg
from reportlab.graphics import renderPM
# import PySimpleGUI as sg
import re
from twocaptcha import TwoCaptcha
import sys
from decouple import config

def captcha_builder(resp):
    with open('../captcha.svg', 'w') as f:
        f.write(re.sub('(<path d=)(.*?)(fill=\"none\"/>)', '', resp['captcha']))

    drawing = svg2rlg('../captcha.svg')
    renderPM.drawToFile(drawing, "captcha.png", fmt="PNG")

    #solved using 2Captcha
    solver = TwoCaptcha(config('API_KEY'))
    try:
        result = solver.normal('captcha.png')

    except Exception as e:
        sys.exit(e)

    else:
        return str(result.get('code'))


    # layout = [[sg.Image('captcha.png')],
    #           [sg.Text("Enter Captcha Below")],
    #           [sg.Input()],
    #           [sg.Button('Submit', bind_return_key=True)]]
    #
    # window = sg.Window('Enter Captcha', layout)
    # event, values = window.read()
    # window.close()
    # return values[1]



src/utils.py
from prettytable import PrettyTable
from captcha import captcha_builder
from datetime import date, datetime, timedelta
import requests
import sys
import hashlib
from decouple import config
import time

resources = {
    'generateMobileOTP': "/api/v2/auth/generateMobileOTP",
    'validateMobileOtp': '/api/v2/auth/validateMobileOtp',
    'beneficiaries': '/api/v2/appointment/beneficiaries',
    'states': '/api/v2/admin/location/states',
    'districts': '/api/v2/admin/location/districts/',
    'schedule': '/api/v2/appointment/schedule',
    'getRecaptcha': '/api/v2/auth/getRecaptcha'
}

coWinUrl = "https://cdn-api.co-vin.in"

header_otp = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36"
}

header = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36"
    , 'content-type': 'application/json'
    , 'origin': 'https://selfregistration.cowin.gov.in'
    , 'sec-fetch-site': 'cross-site'
    , 'sec-fetch-mode': 'cors'
    , 'sec-fetch-dest': 'empty'
    , 'referer': 'https://selfregistration.cowin.gov.in/'
}

beneficiaries_data = {}
states = {}

mobile = 9880267075
secret = "U2FsdGVkX1/VsmHZHLbdwntV6fMy5vTmAZhQtNlj00zdmonoostJjETavz9NKf578AFc3y1bgqAvLExdg48bRA=="
TOKEN_VALIDITY = 840  # taking 1 min buffer


def check_token_status(authentication, header):
    header["Authorization"] = authentication
    url = coWinUrl + resources.get('beneficiaries')
    response = requests.get(url, headers=header)
    if response.status_code == 200:
        return True
    else:
        return False


def generate_otp(mobile, header, secret):
    data = {
        'mobile': mobile,
        'secret': secret
    }
    url = coWinUrl + resources.get("generateMobileOTP")
    response = requests.post(url=url, json=data, headers=header)
    if response.status_code != 200:
        print("Somthing went wrong, regenerating otp")
    else:
        return response.json()


def authenticateOtp(hashed_otp, generate_respose):
    data = {
        'otp': hashed_otp,
        'txnId': generate_respose['txnId']
    }
    url = coWinUrl + resources.get("validateMobileOtp")
    response = requests.post(url=url, json=data, headers=header)
    if response.status_code == 401:
        print("Unauthenticated Access, please check your otp")
    elif response.status_code == 200:
        return response.json()
    else:
        print("Bad Request or Internal Server Error while authentication otp")
        sys.exit(1)


def beneficiaries(authentication, header):
    header["Authorization"] = authentication
    url = coWinUrl + resources.get('beneficiaries')
    response = requests.get(url, headers=header)
    if response.status_code != 200:
        print("Token is not valid, please provide a valid token")
        sys.exit(1)
    else:
        return response.json()


def showBeneficiariesDetails(beneficiaries_response):
    if not beneficiaries_response:
        print("No beneficiaries added")
        return
    table = PrettyTable(['Name', 'Age', 'vaccination_status', 'vaccine'])
    sno = 1
    for item in beneficiaries_response.get('beneficiaries'):
        table.add_row(
            [item.get('name'), (int(datetime.now().year) - int(item.get('birth_year'))), item.get('vaccination_status'),
             item.get('vaccine')])
        beneficiaries_data[str(sno)] = {
            'beneficiary_reference_id': item.get('beneficiary_reference_id'),
            'name': item.get('name'),
            'age': (int(datetime.now().year) - int(item.get('birth_year'))),
            'vaccination_status': item.get('vaccination_status'),
            'vaccine': item.get('vaccine')
        }
        sno = sno + 1
    print(table)
    return beneficiaries_data


def getStates(authentication, header):
    header["Authorization"] = authentication
    url = coWinUrl + resources.get('states')
    response = requests.get(url, headers=header)
    if response.status_code != 200:
        print("Somthing went wrong, regenerating otp")
    else:
        return response.json()


def showStates(states_response):
    table = PrettyTable(['id', 'name'])
    for item in states_response.get('states'):
        table.add_row([item.get('state_id'), item.get('state_name')])
        states[item.get('state_id')] = item.get('state_name')
    print("<========================  STATE NAMES  ============================>")
    print(table)


def getDistricts(state, header):
    url = coWinUrl + resources.get('districts') + state
    response = requests.get(url=url, headers=header)
    if response.status_code != 200:
        print("Somthing went wrong, regenerating otp")
    else:
        return response.json()


def showDistricts(districts_response):
    table = PrettyTable(['id', 'name'])
    for item in districts_response.get('districts'):
        table.add_row([item.get('district_id'), item.get('district_name')])
        states[item.get('state_id')] = item.get('state_name')

    print("<========================  DISTRICT NAMES  ============================>")
    print(table)


def getDateFromUser():
    when_to_book = int(input("1 -> today, 2 -> Tomorrow, Default 2") or "2")
    if when_to_book == 2:
        return (date.today() + timedelta(days=1)).strftime('%d-%m-%Y')
    elif when_to_book == 1:
        return date.today().strftime('%d-%m-%Y')
    else:
        print("Input right value")
        sys.exit(1)


def getSessionsByDistrict(districts, date, header):
    all_sessions = []
    for district in districts:
        session_path = f"/api/v2/appointment/sessions/public/findByDistrict?district_id={district}&date={date}"
        url = coWinUrl + session_path
        response = requests.get(url=url, headers=header)
        if response.status_code != 200:
            print(
                "Somthing went wrong while fetching the sessions, coWin must be under heavy traffic. Please try later!")
            sys.exit(1)
        else:
            curr_dist_session_list = response.json()
            all_sessions.extend(curr_dist_session_list.get('sessions', []))
    return {'sessions': all_sessions}


def getUserConditions():
    fee_type_flag = True
    vaccine_flag = True
    fee_type_preference = int(input("Select the fee type: 1) Free  2) paid, enter) No preference : ") or "0")
    if fee_type_preference == 1:
        fee_type = 'Free'
    elif fee_type_preference == 2:
        fee_type = 'Paid'
    else:
        fee_type = ''
        fee_type_flag = False

    vaccine_type_preference = int(
        input("Select the Vaccine type: 1) Covaxine  2) Covishield, enter) No preference : ") or "0")
    if vaccine_type_preference == 1:
        vaccine_type = 'COVAXIN'
    elif vaccine_type_preference == 2:
        vaccine_type = 'COVISHIELD'
    else:
        vaccine_type = ''
        vaccine_flag = False

    dose_number_preference = int(
        input("Is this your 1st dose or 2nd: 1) dos1  2) dos2, enter) default dos1 : ") or "1")

    return fee_type, fee_type_flag, vaccine_type, vaccine_flag, dose_number_preference


def generate_captcha(request_header):
    print('================================= GETTING CAPTCHA ==================================================')
    CAPTCHA_URL = coWinUrl + resources.get('getRecaptcha')
    resp = requests.post(CAPTCHA_URL, headers=request_header)
    print(f'Captcha Response Code: {resp.status_code}')

    if resp.status_code == 200:
        return captcha_builder(resp.json())


def book_slot(header, mini_slot, session_id, slot, beneficiary_reference_ids, center_id):
    url = coWinUrl + resources.get('schedule')
    data = {
        'center_id': center_id,
        'dose': mini_slot,
        'session_id': session_id,
        'slot': slot,
        'beneficiaries': beneficiary_reference_ids
    }
    # captcha = generate_captcha(request_header=header)
    # data['captcha'] = captcha
    print("Requested data to book slot: ", data)
    return requests.post(url=url, json=data, headers=header)

def generate_and_validate_otp(mobile, header_otp, secret, auto):
    # generate otp
    global otp_message
    generate_respose = generate_otp(mobile, header_otp, secret)
    print("generate_respose recevied: ", generate_respose)
    # enter Otp
    if not auto:
        otp = input("enter the otp: ")
    else:
        print("Fetching otp after 10secs")
        time.sleep(10)
        i = 1
        while i <= 6:
            kvdb_url = config('KVDB_URL') + '/' + str(mobile)
            otp_message_response = requests.get(kvdb_url)
            otp_message = otp_message_response.text
            print("otp_message output:", otp_message)
            if len(otp_message) == 0:
                i = i + 1
                time.sleep(5)
            else:
                i = 999
        if i == 7:
            print("No otp received from kvdb.io, please try again")
            sys.exit(1)
        otp = otp_message[37:43]
    hashed_otp = hashlib.sha256(str(otp).encode('UTF-8')).hexdigest()
    token_json = authenticateOtp(hashed_otp, generate_respose)
    print("Token recevied: ", token_json)
    # collect Barer token to check validity of the session
    return token_json.get("token")